File: built-in plugins dsp/mb-drums-electronic.cpp
/**
 * MB Electronic Kit
 * Category : instrument
 * Type     : drums
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic 808/909 electronic drums
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DRUMS_ELECTRONIC_H
#define MB_DRUMS_ELECTRONIC_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDrumsElectronic : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-drums-electronic";
    static constexpr const char* PLUGIN_NAME    = "MB Electronic Kit";
    static constexpr const char* PLUGIN_TYPE    = "drums";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float decay = 0.5f;  // range [0.1, 2]
    float volume = 0.8f;  // range [0, 1]
    };

    MbDrumsElectronic() = default;
    ~MbDrumsElectronic() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.decay = std::clamp(params.decay, 0.1f, 2f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Electronic Kit
        return input;
    }
};

#endif // MB_DRUMS_ELECTRONIC_H

File: built-in plugins dsp/mb-drums-electronic.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbDrumsElectronicPlugin: PluginDefinition = {
  id: "mb-drums-electronic",
  slug: "mb-drums-electronic",
  name: "MB Electronic Kit",
  category: "instrument",
  type: "drums",
  version: "1.0.0",
  description: "Classic 808/909 electronic drums",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sine", detune: 0, gain: 0.8 },
    { type: "noise", detune: 0, gain: 0.6 },
  ],
  envelope: { attack: 0.001, decay: 0.3, sustain: 0.0, release: 0.4 },
  parameters: [
    {
      id: "decay",
      name: "Decay",
      type: "float",
      defaultValue: 0.5,
      minValue: 0.1,
      maxValue: 2,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { decay: 0.5, volume: 0.8 },
};

export default MbDrumsElectronicPlugin;
